jib-core/src/main/java/com/google/cloud/tools/jib/event/progress/ThrottledConsumer.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.event.progress;
+
+import java.io.Closeable;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.function.BinaryOperator;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+/**
+ * Wraps a {@link Consumer} so that multiple consume calls ({@link #accept}) within a short period
+ * of time are merged into a single later call.
+ */
+public class ThrottledConsumer<T> implements Consumer<T>, Closeable {
```

Comment:
Can you kind of explain why this is needed? Are we doing something unavoidable somewhere that requires this? Maybe I missed the thread where this was already discussed.

Reply:
Currently, this is only used for throttling progress event dispatch. `ListenableCountingOutputStream` had the throttling capability integrated for listening, and in this PR, I factored out the throttling functionality into this independent class. I am pretty sure throttling is not strictly necessary, but perhaps Q wanted to avoid dispatching too many progress update in a very short time span. For example, if debug logging is enabled, we will see too many progress events, as I guess `ListenableCountingOutputStream` may call `write()` insanely many times for a large blob push/pull.

Reply:
(I had added a note to this line in the original commit that explained this is factored out from another place, but I see the note got removed when I renamed this class.)

Reply:
What do you think about a minimal class (without the generics, injected adders, etc) that just handles the progress event consumption? This class seems a little over-engineered for what we want it to do?

Reply:
Yeah, I think we can specialize it. No need to be overly generic.